core/storage/pager.rs

```diff
@@ -1345,63 +1380,249 @@ impl Pager {
+                    #[cfg(not(feature = "omit_autovacuum"))]
+                    let mut new_db_size = old_db_size;
+                    #[cfg(feature = "omit_autovacuum")]
+                    let new_db_size = old_db_size;
+
+                    tracing::debug!("allocate_page(database_size={})", new_db_size);
+                    #[cfg(not(feature = "omit_autovacuum"))]
+                    {
+                        //  If the following conditions are met, allocate a pointer map page, add to cache and increment the database size
+                        //  - autovacuum is enabled
+                        //  - the last page is a pointer map page
+                        if matches!(*self.auto_vacuum_mode.borrow(), AutoVacuumMode::Full)
+                            && is_ptrmap_page(
+                                new_db_size + 1,
+                                header_accessor::get_page_size(self)? as usize,
+                            )
+                        {
+                            // we will allocate a ptrmap page, so increment size
+                            new_db_size += 1;
+                            let page =
+                                allocate_new_page(new_db_size as usize, &self.buffer_pool, 0);
+                            self.add_dirty(&page);
+                            let page_key = PageCacheKey::new(page.get().id);
+                            let mut cache = self.page_cache.write();
+                            match cache.insert(page_key, page.clone()) {
+                                Ok(_) => (),
+                                Err(CacheError::Full) => return Err(LimboError::CacheFull),
+                                Err(_) => {
+                                    return Err(LimboError::InternalError(
+                                        "Unknown error inserting page to cache".into(),
+                                    ))
+                                }
+                            }
+                        }
+                    }
 
-        #[cfg(not(feature = "omit_autovacuum"))]
-        {
-            //  If the following conditions are met, allocate a pointer map page, add to cache and increment the database size
-            //  - autovacuum is enabled
-            //  - the last page is a pointer map page
-            if matches!(*self.auto_vacuum_mode.borrow(), AutoVacuumMode::Full)
-                && is_ptrmap_page(new_db_size, header_accessor::get_page_size(self)? as usize)
-            {
-                let page = allocate_page(new_db_size as usize, &self.buffer_pool, 0);
-                self.add_dirty(&page);
+                    let first_freelist_trunk_page_id =
+                        header_accessor::get_freelist_trunk_page(self)?;
+                    if first_freelist_trunk_page_id == 0 {
+                        *state = AllocatePageState::AllocateNewPage {
+                            current_db_size: new_db_size,
+                        };
```

Comment:
This doesn't need to be a state, it adds unnecessary loop work. Maybe create `allocate_page_without_freelist` and call it there.